src/decode.rs

```diff
@@ -4830,48 +4901,48 @@ pub(crate) fn rav1d_decode_frame_exit(
     // We use a blocking lock here because we have rare contention with other
     // threads.
     let mut f = fc.data.write();
-    if f.sr_cur.p.data.is_some() {
+    if f.content.sr_cur.p.data.is_some() {
         task_thread.error.store(0, Ordering::Relaxed);
     }
-    let cf = f.frame_thread.cf.get_mut();
+    let cf = f.content.frame_thread.cf.get_mut();
     if c.fc.len() > 1 && retval.is_err() {
         cf.fill_with(Default::default);
     }
 
     if retval.is_ok() && c.fc.len() > 1 && c.strict_std_compliance {
-        if f.refp.iter().any(|rf| {
+        if f.content.refp.iter().any(|rf| {
             rf.p.frame_hdr.is_some()
                 && rf.progress.as_ref().unwrap()[1].load(Ordering::SeqCst) == FRAME_ERROR
         }) {
             retval = Err(EINVAL);
             task_thread.error.store(1, Ordering::SeqCst);
-            f.sr_cur.progress.as_mut().unwrap()[1].store(FRAME_ERROR, Ordering::SeqCst);
+            f.content.sr_cur.progress.as_mut().unwrap()[1].store(FRAME_ERROR, Ordering::SeqCst);
         }
     }
 
-    let _ = mem::take(&mut f.refp);
-    let _ = mem::take(&mut f.ref_mvs);
-    let _ = mem::take(&mut f.cur);
-    let _ = mem::take(&mut f.sr_cur);
+    let _ = mem::take(&mut f.content.refp);
+    let _ = mem::take(&mut f.content.ref_mvs);
+    let _ = mem::take(&mut f.content.cur);
+    let _ = mem::take(&mut f.content.sr_cur);
     let _ = mem::take(&mut *fc.in_cdf.try_write().unwrap());
     if let Some(frame_hdr) = &f.frame_hdr {
         if frame_hdr.refresh_context != 0 {
-            if let Some(progress) = f.out_cdf.progress() {
+            if let Some(progress) = f.content.out_cdf.progress() {
                 progress.store(
                     if retval.is_ok() { 1 } else { TILE_ERROR as u32 },
                     Ordering::SeqCst,
                 );
             }
-            let _ = mem::take(&mut f.out_cdf);
+            let _ = mem::take(&mut f.content.out_cdf);
         }
     }
 
-    let _ = mem::take(&mut f.cur_segmap);
-    let _ = mem::take(&mut f.prev_segmap);
-    let _ = mem::take(&mut f.mvs);
+    let _ = mem::take(&mut f.content.cur_segmap);
+    let _ = mem::take(&mut f.content.prev_segmap);
+    let _ = mem::take(&mut f.content.mvs);
     let _ = mem::take(&mut f.seq_hdr);
     let _ = mem::take(&mut f.frame_hdr);
```

Comment:
any time the `seq_hdr`/`frame_hdr` fields are `take`n from like this, it would be absolutely critical to reconstruct the `Arc<T>` if these were `Some` so as to not leak the header refcount. seems pretty subtle and easy to mistake. might be less bad with a `Rav1dFrameDataMaybeData::drop_state()` or something, but ouch.